sdks/python/apache_beam/typehints/typecheck.py

```diff
@@ -265,3 +269,95 @@ def visit_transform(self, applied_transform):
                 transform.get_type_hints(),
                 applied_transform.full_label),
             applied_transform.full_label)
+
+
+class PerformanceTypeCheckVisitor(pipeline.PipelineVisitor):
+
+  _in_combine = False
+  combine_classes = (
+      core.CombineFn,
+      core.CombinePerKey,
+      core.CombineValuesDoFn,
+      core.CombineValues,
+      core.CombineGlobally)
+
+  def enter_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = True
+
+  def leave_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, self.combine_classes):
+      self._in_combine = False
+
+  def visit_transform(self, applied_transform):
+    transform = applied_transform.transform
+    if isinstance(transform, core.ParDo) and not self._in_combine:
+      # Store output type hints in current transform
+      transform.fn._runtime_output_constraints = {
+          applied_transform.full_label: self.get_output_type_hints(transform)
+      }
+
+      # Store input type hints in producer transform
+      producer = applied_transform.inputs[0].producer
+      if (hasattr(producer, 'transform') and
+          hasattr(producer.transform, 'fn') and
+          hasattr(producer.transform.fn, '_runtime_output_constraints')):
+        producer = producer.transform.fn
+        producer._runtime_output_constraints[applied_transform.full_label] \
```

Comment:
Style: in Beam we avoid backslash line brakes. You can use yapf to break your lines for you. (If needed, add `()`'s)>